torch_xla/debug/metrics_compare_utils.py

```diff
@@ -0,0 +1,180 @@
+
+  Returns:
+    dict of metric name to list of values, one value from each report in
+    metrics_reports. Order of values for each metric in the output dict
+    will match the order of metrics reports in the input.
+    
+    For example, if the args were [report1, report2, report3], output might be:
+    {
+      'CompileTime__Accumulator_sec': [50, None, 80],  # missing from report2
+      'InboundData__TotalSamples': [11, 12, 11],
+      'CreateCompileHandles__Value': [20, 20, 19],
+      ...
+    }
+  """
+  data_points = collections.defaultdict(lambda: [None] * len(metrics_reports))
+  for report_index in range(len(metrics_reports)):
+    parsed_report = _parse_metrics_report(metrics_reports[report_index])
+    for metric_name in parsed_report:
+      data_points[metric_name][report_index] = parsed_report.get(
+          metric_name, None)
+  return data_points
+
+
+def _compute_aggregates(data_points):
+  # Compute mean and standard deviation for each key in data_points.
+  aggregates = {}
+  for k, v_list in data_points.items():
+    v_list = [v for v in v_list if v is not None]
+    if v_list:
+      aggregates[k] = {'mean': mean(v_list), 'stddev': std(v_list)}
+  return aggregates
+
+
+def compare_metrics(data_points, metrics_report,
+                    config={'base_tolerance': 2.0}):
+  """Compare metrics_report to historical averages and report differences.
+
+  Args:
+    data_points(dict): Output of get_data_points_from_metrics_reports().
+    metrics_report(string): Output of torch_xla metrics_report(). These will
+      be compared with the aggregates of raw_data_points.
+    config(dict): Configuration for metrics comparison. Tolerances are
+      expressed as the number of standard deviations away from the mean.
+      You can supply a per-metric tolerance for any aggregated metric name,
+      otherwise base_tolerance will be used. Example config:
+      {
+        'base_tolerance': 2.0,
+        'TransferFromServerTime_Accumulator_sec_tolerance': 10.0,
+        ...
+      }
+
+  Returns:
+    Metrics difference report (string). For any metric that differed between
+    metrics_report and the aggregates of raw_data_points, this report will have
+    1 line reporting the difference.
+  """
+  parsed_report = _parse_metrics_report(metrics_report)
+  means_and_stddevs = _compute_aggregates(data_points)
+
+  difference_report = ''
+  for k, v in sorted(parsed_report.items()):
+    if k not in means_and_stddevs:
+      # Alert if we have new ops falling back to CPU.
+      if k.startswith('aten::'):
+        difference_report += 'Found new aten counter: {}: {}\n'.format(k, v)
+    else:
+      v_mean = means_and_stddevs[k]['mean']
+      v_stddev = means_and_stddevs[k]['stddev']
+
+      # base_tolerance can be overridden for individual metrics.
+      mul = config.get('{}_tolerance'.format(k), None)
+      if mul is None:
+        mul = config['base_tolerance']
+
+      lower_limit = v_mean - (v_stddev * mul)
```

Comment:
I think we can do a bit more flexible here.
What we have in the TPU dashboard/monitoring is the ability to specify python expressions, which we then `eval()` with a dictionary built with the metrics inputs.

Also, below, the bounding up/down is not good.
Example, if we improve the step time a lot, should that trigger an alert?

Reply:
up/down: Initially I wanted to catch any suspicious activity and also not assume that lower is better for every metric, but I agree that right now lower actually is better for every metric I see.

I removed the lower bound check and added functionality to use a custom expression. That way, if we do end up with a metric where higher is better, we can just pass the expression there. Updated the docstring to explain usage